feat: introduce tests

ffMathy opened the pull request:
<!--
  [For internal use] 
  Copy the task id from the bottom of the sidebar and paste it after FTRACK-

  Resolves FTRACK-

-->

- [ ] I have added automatic tests where applicable
- [ ] The PR title is suitable as a release note
- [ ] The PR contains a description of what has been changed
- [ ] The description contains manual test instructions

## Changes

<!--
  What are you changing and what is the reason? If there are any UI changes, include a screenshot for the changes.
-->

## Test

<!--
  How should this be tested? Include any requirements on other repositories or environment.
  For bug fixes, include the original reproduction steps.
-->

ffMathy committed:
refactor: break out generate file

ffMathy committed:
feat: add vitest

ffMathy committed:
lint: fix linting issues

ffMathy commented:
Hey @gismya here's my plan for this PR and upcoming ones.

1. Introducing tests.
2. I'll be switching to TS-SDK for emitting (fixing #8).
3. I'll be introducing support for types and object types, and having a list of those available at runtime.

This is the first PR. In that regard, I have some questions:

1. Is it correct you want to use `vitest` for tests still?
2. Can you elaborate on why `QuerySchemasResponse[]` is used? So in essence, there's an array of arrays of schemas? What does this structure mean?

ffMathy committed:
test: add basic test

gismya commented:
> Hey @gismya here's my plan for this PR and upcoming ones.
> 
> 1. Introducing tests.
> 2. I'll be switching to TS-SDK for emitting (fixing [feat: use TypeScript's emit SDK #8](https://github.com/ftrackhq/ftrack-ts-schema-generator/issues/8)).
> 3. I'll be introducing support for types and object types, and having a list of those available at runtime.

Tests are always good. With only a cursory look, utilizing the TS-SDK seems a bit more complicated to use and maintain than just string concatenation. What are the main benefits of doing it? I wouldn't want you spending time and effort on something we later find too complex to merge or something like that. The complicated of the current code, in my mind, is the handling of the schemas, not the actual creation of the TS file?

Could you elaborate on point 3?



> This is the first PR. In that regard, I have some questions:
> 
> 1. Is it correct you want to use `vitest` for tests still?
> 2. Can you elaborate on why `QuerySchemasResponse[]` is used? So in essence, there's an array of arrays of schemas? What does this structure mean?

Yes, vitest is the preferred testing solution.
`QuerySchemasResponse[]` is because that's what's returned from the session call in `getSchemas`, since session.call can take multiple calls at the same time. Then that first layer of array should have been flattened, but I missed that when writing the initial implementation and didn't really see it until you asked now :)

ffMathy commented:
Thanks for elaborating! 😍

As for the tssdk, I wouldn't be worried. Get what you mean, but I don't think it'll be too bad. And I'll make it as a separate PR afterwards 😊🙏

ffMathy commented:
As for the benefits of it, the main benefit is higher readability 😊

ffMathy commented:
As for point 3, I'd love if the object types and types in ftrack were emitted as a string union type.

ffMathy committed:
fix: add more snapshot tests

ffMathy commented:
This is now ready for review! I ended up going for snapshot testing in vitest. I think this is a good match for types of tests like these.

Not sure if you've used it before, but if a snapshot fails, it shows a diff, and then provides you with an option to accept the new shapshot.

That way, it becomes clear what effects a change had. Works well with watch mode too.

ffMathy committed:
fix: dynamic copyright year

ffMathy commented on source/convertSchemaToInterface.ts:
```diff
@@ -166,7 +167,8 @@ function convertTypeToTsType(key: string, value?: TypedSchemaProperty) {
       return `${value.items.$ref}[]`;
     } else if ("type" in value.items && value.items.type) {
-      return `${convertTypeToTsType(value.items.type as string)}[]`;
+      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
```
This was a bit weird...

In the call below, the second argument wasn't specified, and hence would default to `undefined`. But that should crash the recursive call, since we (on line 156 check for the type of the value).

So I actually think we have a runtime bug here, but not sure how to fix it, or how to test it. Not sure what the intended behavior is.

If you can tell me, I can update accordingly, add a test, and fix the bug.

gismya replied:
Mistake on my part. That never happens. There are properties that are named `items`, and I must have read the name of the property instead of the `items` key inside that property and confused myself.

ffMathy replied:
So do I just remove the whole `else if`?

gismya replied:
Yes

gismya replied:
Great that the tests are set up, because there shouldn't be any difference :)

ffMathy replied:
Haha yeah wow didn't even think about that 🤣 the full schema snapshot test should be able to verify it. Thanks!

ffMathy replied:
This has been fixed now.

ffMathy committed:
fix: linting

ffMathy committed:
fix: various fixes

- separate files for each snapshot.
- bugfix for aliases.
- added default schema test for newly created Ftrack trial instance.

ffMathy committed:
fix: linting

ffMathy committed:
refactor: rename response files

gismya commented:
> As for the benefits of it, the main benefit is higher readability 😊

If you manage to make it more readable I'm all for it. But I think the parsing part is the confusing part, not the actual writing to the schema. But I'd love to be wrong :)

ffMathy commented:
> > As for the benefits of it, the main benefit is higher readability 😊
> 
> If you manage to make it more readable I'm all for it. But I think the parsing part is the confusing part, not the actual writing to the schema. But I'd love to be wrong :)

Yeah on that front I am also hoping for contributions.

But either way, I know it's opinionated and risky, and there'll be no hard feelings if the PR is rejected. I'm having fun working on it either way.

What do you think about the PR from a review perspective?

gismya commented:
> What do you think about the PR from a review perspective?

I don't know. :) I'll try to make time for reviewing it later today.

gismya commented on source/convertSchemaToInterface.ts:
```diff
@@ -124,9 +125,9 @@ function convertPropertiesToTypes(
     }
     // If neither type or $ref is defined, we can't generate a type. Log an error
-    if (!("type" in value) && !("$ref" in value)) {
+    if (!("type" in value) && !("$ref" in value) && !("alias_for" in value)) {
```
What's this change for?

ffMathy replied:
Actually, I was hoping you could tell me, and if it's correctly made.

Without the code, the test for the full schema fails. If I recall correctly, something about a type property missing or something on some of the schemas.

ffMathy replied:
I just tried removing it now. The test fails with:

```
FAIL  source/emit.test.ts > default ftrack schema
AssertionError: expected [ Array(1) ] to deeply equal []

- Expected
+ Received

- Array []
+ Array [
+   "No type or $ref defined for property user_id in schema UserView",
+ ]
```

gismya replied:
I patched the file and removed the `alias_for` check

ffMathy replied:
Thank you for that!

gismya commented on source/emit.test.ts:
```diff
@@ -0,0 +1,237 @@
+
+test("schema subtype of TypedContext", async () => {
+  //arrange
+  vi.setSystemTime(new Date(2023, 1, 1));
```
Might be easier to add this to beforeEach

ffMathy replied:
Yeah I did consider this, but one of the tests (and possibly more in the future) are not snapshot tests. Let me know if you still think I should add it anyway, and I'll do it 🙏

ffMathy replied:
Actually, at second thought, this is just in general a good idea. I did it now.

gismya reviewed (changes requested):
Looks great. One optional suggestion and one change need to understand what it's for, then it's good to merge

ffMathy committed:
fix: review fixes

ffMathy committed:
lint: prettier

gismya committed:
Fix: Incorrect schema response file

ffMathy commented on source/__snapshots__/responses/query_schemas.json:
```diff
@@ -3989,6 +3989,7 @@
         "type": "boolean"
       },
       "user_id": {
+        "type": "string",
```
Thanks!

ffMathy commented:
Thanks! I corrected everything now.

Do we merge now, or do we wait until that Ftrack schema bug is fixed, so the tests pass?

gismya commented:
You can merge now

ffMathy commented:
Cool! Do I have permissions to merge it? Doesn't look like it 🫣

gismya commented:
Right,  forgot about that detail
